adds a new guide on mastering partitioned tables in TiDB (#21846)
pingcap/docs: best-practices/tidb-partitioned-tables-best-practices.md

---
title: Best Practices for Using TiDB Partitioned Tables
summary: Learn best practices for using TiDB partitioned tables to improve performance, simplify data management, and handle large-scale datasets efficiently.
---

# Best Practices for Using TiDB Partitioned Tables

This guide introduces how to use partitioned tables in TiDB to improve performance, simplify data management, and handle large-scale datasets efficiently.

> Partitioned tables in TiDB offer a versatile approach to managing large datasets, improving query efficiency, facilitating bulk data deletion, and alleviating write hotspot issues. By dividing data into logical segments, TiDB can leverage partition pruning to skip irrelevant data during query execution, reducing resource consumption and accelerating performance, particularly in Online Analytical Processing (OLAP) workloads with massive datasets.

Reviewer: ```suggestion
This guide describes how to use partitioned tables in TiDB to improve performance, simplify data management, and handle large-scale datasets efficiently.

Partitioned tables in TiDB provide a versatile approach to managing large datasets, improving query efficiency, facilitating bulk data deletion, and alleviating write hotspot issues. By dividing data into logical segments, TiDB can leverage partition pruning to skip irrelevant data during query execution. This reduces resource consumption and improves performance, particularly in Online Analytical Processing (OLAP) workloads with large datasets.
```

Although partitioning provides clear benefits, it also introduces challenges. For example, newly created Range partitions can create temporary hotspots. To address this issue, TiDB supports automatic or manual Region pre-splitting to balance data distribution and avoid bottlenecks.

This document examines partitioned tables in TiDB from several perspectives, including query optimization, data cleanup, write scalability, and index management. It also provides practical guidance on how to optimize partitioned table design and tune performance in TiDB through detailed scenarios and best practices.

> **Note:** 
>
> To get started with the fundamentals, see [Partitioning](/partitioned-table.md), which explains key concepts such as partition pruning, index types, and partitioning methods.

## Improve query efficiency

This section describes how to improve query efficiency by the following methods:

- Partition pruning

> - Query performance on secondary indexes

Reviewer: ```suggestion
- [Partition pruning](#partition-pruning)
- [Query performance on secondary indexes](#query-performance-on-secondary-indexes-non-partitioned-tables-vs-local-indexes-vs-global-indexes)
```

### Partition pruning

> Partition pruning is an optimization technique that allows TiDB to reduce the amount of data scanned when executing queries against partitioned tables. Instead of scanning all partitions, TiDB analyzes filter conditions of the query and determines which partitions might contain relevant data, and scans only those partitions. This significantly improves query performance by reducing I/O and computation overhead.

Reviewer: ```suggestion
Partition pruning is an optimization technique that reduces the amount of data TiDB scans when querying partitioned tables. Instead of scanning all partitions, TiDB evaluates the query filter conditions to identify the partitions that might contain matching data and scans only those partitions. This approach reduces I/O and computation overhead, which significantly improves query performance.
```

> Partition pruning is most beneficial in scenarios where query predicates match the partitioning strategy. Common use cases include:

Reviewer: ```suggestion
Partition pruning is most effective when query predicates align with the partitioning strategy. Typical use cases include the following:
```

- Time-series data queries: when data is partitioned by time ranges (for example, daily, monthly), queries restricted to a specific time period can quickly skip unrelated partitions.
- Multi-tenant or category-based datasets: partitioning by tenant ID or category enables queries to focus on a small subset of partitions.

> - Hybrid Transactional and Analytical Processing (HTAP): especially for range partitioning, TiDB can leverage partition pruning in analytical workloads on TiFlash to skip irrelevant partitions and scan only the necessary subset, preventing full table scans on large datasets.

Reviewer: ```suggestion
- Hybrid Transactional and Analytical Processing (HTAP): especially for Range partitioning, TiDB can apply partition pruning to analytical workloads on TiFlash. This optimization skips irrelevant partitions and avoids full table scans on large datasets.
```

**With hash or key partitioning:**

- The table and the secondary indexes are split into multiple partitions using a hash or key function on the primary key or indexed columns.
- Each partition has its own set of Regions, often distributed across different TiKV nodes.
- Inserts are spread across multiple Regions in parallel, improving workload distribution and throughput.

### Use cases

If a table with an [`AUTO_INCREMENT`](/auto-increment.md) primary key experiences heavy bulk inserts and suffers from write hotspot issues, applying hash or key partitioning on the primary key can help distribute the write workload more evenly.

```sql
CREATE TABLE server_info (
  id bigint NOT NULL AUTO_INCREMENT,
  serial_no varchar(100) DEFAULT NULL,
  device_name varchar(256) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci DEFAULT NULL,
  device_type varchar(50) DEFAULT NULL,
  modified_ts timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
  PRIMARY KEY (id) /*T![clustered_index] CLUSTERED */,
  KEY idx_serial_no (serial_no),
  KEY idx_modified_ts (modified_ts)
) /*T![auto_id_cache] AUTO_ID_CACHE=1 */
PARTITION BY KEY (id) PARTITIONS 16;
```

> ### Pros

Reviewer: ```suggestion
### Advantages
```
https://developers.google.com/style/word-list#pros
Don't use. Instead, use a more precise term, such as advantages.

- Balanced write workload: hotspots are spread across multiple partitions, and therefore multiple Regions, reducing contention and improving insert performance.
- Query optimization via partition pruning: if queries already filter by the partition key, TiDB can prune unused partitions, scanning less data and improving query speed.

> ### Cons

Reviewer: ```suggestion
### Disadvantages
```
https://developers.google.com/style/word-list#cons
Don't use. Instead, use a more precise term, such as disadvantages.

```sql
ALTER TABLE employees ADD PARTITION (PARTITION p4 VALUES LESS THAN (2011));

SHOW TABLE employees PARTITION (p4) regions;

SPLIT PARTITION TABLE employees INDEX `PRIMARY` BETWEEN (1, "2006-01-01") AND (100000, "2011-01-01") REGIONS <number_of_regions>;

SPLIT PARTITION TABLE employees PARTITION (p4) INDEX `idx_employees_on_store_id` BETWEEN (1) AND (1000) REGIONS <number_of_regions>;

SHOW TABLE employees PARTITION (p4) regions;
```

### Solutions for clustered partitioned tables

#### Pros

Queries using **Point Get** or **Table Range Scan** do not need additional lookups, resulting in better read performance.

#### Cons

Manual region splitting is required when creating new partitions, increasing operational complexity.

#### Suitable scenarios

It is suitable when low-latency point queries are important and operational resources are available to manage region splitting.

#### Best practices

To address hotspot issues caused by new range partitions, you can perform the steps described in [Best practices for non-clustered partitioned tables](#best-practices).

### Solutions for clustered non-partitioned tables

> #### Pros

Reviewer: ```suggestion
#### Advantages
```

(PARTITION `fa_2024001` VALUES LESS THAN (2025001),
PARTITION `fa_2024002` VALUES LESS THAN (2025002),
PARTITION `fa_2024003` VALUES LESS THAN (2025003),
...
...
PARTITION `fa_2024365` VALUES LESS THAN (2025365));
```

### Table schema for a non-partitioned table: `fa_new`

```sql
CREATE TABLE `fa_new` (
  `id` bigint NOT NULL AUTO_INCREMENT,
  `account_id` bigint(20) NOT NULL,
  `sid` bigint(20) DEFAULT NULL,
  `user_id` bigint NOT NULL,
  `date` int NOT NULL,
  PRIMARY KEY (`id`,`date`) /*T![clustered_index] CLUSTERED */,
  KEY `index_fa_on_sid` (`sid`),
  KEY `index_fa_on_account_id` (`account_id`),
  KEY `index_fa_on_user_id` (`user_id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
```

These examples show converting a partitioned table to a non-partitioned table, but the same methods also work for converting a non-partitioned table to a partitioned table.

### Method 1: Pipelined DML `INSERT INTO ... SELECT`

```sql
SET tidb_dml_type = "bulk";
SET tidb_mem_quota_query = 0;
SET tidb_enable_mutation_checker = OFF;
INSERT INTO fa_new SELECT * FROM fa;
-- 120 million rows copied in 58m 42s
```

### Method 2: `IMPORT INTO ... FROM SELECT`

> 

Reviewer 2: ```suggestion
```

- Hybrid Transactional and Analytical Processing (HTAP): especially for Range partitioning, TiDB can apply partition pruning to analytical workloads on TiFlash. This optimization skips irrelevant partitions and avoids full table scans on large datasets.

For more use cases, see [Partition Pruning](/partition-pruning.md).

### Query performance on secondary indexes: non-partitioned tables vs. local indexes vs. global indexes

In TiDB, partitioned tables use local indexes by default, where each partition maintains its own set of indexes. In contrast, a global index covers the entire table in one index and tracks rows across all partitions.

For queries that access data from multiple partitions, global indexes generally provide better performance. This is because a query using local indexes requires separate index lookups in each relevant partition, while a query using a global index performs a single lookup across the entire table.

#### Types of tables to be tested

The query performance of the following types of tables are evaluated:

- Non-partitioned tables
- Partitioned tables with global indexes
- Partitioned tables with local indexes

#### Test setup

> 

Reviewer: ```suggestion

The test uses the following configuration:

```

- The partitioned table has 365 partitions, defined by the range partitioning on a date column.
- Each matching key returns multiple rows, simulating a high-volume OLTP-style query pattern.
- The impact of different partition counts is also evaluated to understand how partition granularity influences latency and index performance.

#### Schema

The following schema is used in the example.

```sql
CREATE TABLE `fa` (
  `id` bigint NOT NULL AUTO_INCREMENT,
  `account_id` bigint(20) NOT NULL,
  `sid` bigint(20) DEFAULT NULL,
  `user_id` bigint NOT NULL,
  `date` int NOT NULL,
  PRIMARY KEY (`id`,`date`) /*T![clustered_index] CLUSTERED */,
  KEY `index_fa_on_sid` (`sid`),
  KEY `index_fa_on_account_id` (`account_id`),
  KEY `index_fa_on_user_id` (`user_id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
PARTITION BY RANGE (`date`)
(PARTITION `fa_2024001` VALUES LESS THAN (2025001),
PARTITION `fa_2024002` VALUES LESS THAN (2025002),
PARTITION `fa_2024003` VALUES LESS THAN (2025003),
...
...

> PARTITION `fa_2024365` VALUES LESS THAN (2025365));

Reviewer: ```suggestion
PARTITION BY RANGE (`date`)(
  PARTITION `fa_2024001` VALUES LESS THAN (2025001),
  PARTITION `fa_2024002` VALUES LESS THAN (2025002),
  PARTITION `fa_2024003` VALUES LESS THAN (2025003),
  ...
  PARTITION `fa_2024365` VALUES LESS THAN (2025365)
);
```

> PARTITION `fa_2024365` VALUES LESS THAN (2025365));

Reviewer 3: Oh, I already asked this before...

```

#### SQL

The following SQL statement is used in the example.

```sql
SELECT `fa`.*
FROM `fa`
WHERE `fa`.`sid` IN (
  1696271179344,
  1696317134004,
  1696181972136,
  ...
  1696159221765
);
```

- Query filters on secondary index, but does not include the partition key.
- Causes local indexes key lookup for each partition due to lack of pruning.
- Table lookup tasks are significantly higher for partitioned tables.

#### Test results

The following table shows the test results.

| Configuration | Average Query Time | Cop task for index range scan | Cop task for table lookup | Total Cop tasks | Key Takeaways |
|---|---|---|---|---|---|
| Non-partitioned table | 12.6 ms | 72 | 79 | 151 | Provides the best performance with the fewest Cop tasks, which is ideal for most OLTP use cases. |
| Partitioned table with local indexes | 108 ms | 600 | 375 | 975 | When the partition key is not used in the query condition, local index queries scan all partitions. |

> | Partitioned table with global indexes | 14.8 ms | 69 | 383 | 452 | It improves index scan efficiency, but table lookups can still take a long time if many rows match. |

Reviewer: ```suggestion
| Configuration | Average query time | Cop tasks (index scan) | Cop tasks (table lookup) | Total Cop tasks |
|---|---|---|---|---|
| Non-partitioned table | 12.6 ms | 72 | 79 | 151 |
| Partitioned table with local indexes | 108 ms | 600 | 375 | 975 |
| Partitioned table with global indexes | 14.8 ms | 69 | 383 | 452 |

- **Non-partitioned table**: provides the best performance with the fewest tasks. Suitable for most OLTP workloads.
- **Partitioned table with global indexes**: improve index scan efficiency, but table lookups remain expensive when many rows match.
- **Partitioned table with local indexes**: when the query condition does not include the partition key, local index queries scan all partitions.
```

Move key takeaways out of the table and into a bulleted list. Wide tables are difficult to read on the website.

Data comes from a table with 365 range partitions (for example, by date).

- The **Average Query Time** is obtained from the `statement_summary` view.

> - The query uses a secondary index and returns 400 rows.

Reviewer: ```suggestion
```
Consolidate the context into a single introductory sentence in L110. The current format separates related information across a paragraph and bullet points, which is harder to scan.

| TableRowIDScan_6(Probe)   | 398.73  | 166072.78 | 400     | cop[tikv] | table:fa                             | time:7.01ms, loops:2, cop_task:{num:79, max:4.98ms, min:0s, avg:514.9µs, p95:3.75ms, max_proc_keys:10, p95_proc_keys:5, tot_proc:15ms, tot_wait:21.4ms, copr_cache_hit_ratio:0.00, build_task_duration:341.2µs, max_distsql_concurrency:1, max_extra_concurrency:7, store_batch_num:62}, rpc_info:{Cop:{num_rpc:17, total_time:40.5ms}}, tikv_task:{proc max:0s, min:0s, avg:0s, p80:0s, p95:0s, iters:79, tasks:79}, scan_detail:{total_process_keys:400, total_process_keys_size:489856, total_keys:800, get_snapshot_time:20.8ms, rocksdb:{key_skipped_count:400, block:{cache_hit_count:1600}}}, time_detail:{total_process_time:15ms, total_wait_time:21.4ms, tikv_wall_time:10.9ms} | keep order:false | N/A | N/A |
```

> The following is an execution plan example for a partitioned tables with a global index:

Reviewer: ```suggestion
</details>

<details>
<summary><b>Partitioned table with global indexes</b></summary>
```

| TableRowIDScan_6(Probe)| 398.73  | 165221.49 | 400     | cop[tikv] | table:fa                             | time:514ms, loops:434, cop_task:{num:375, max:31.6ms, min:0s, avg:1.33ms, p95:1.67ms, max_proc_keys:2, p95_proc_keys:2, tot_proc:220.7ms, tot_wait:242.2ms, copr_cache_hit_ratio:0.00, build_task_duration:27.8ms, max_distsql_concurrency:1, max_extra_concurrency:1, store_batch_num:69}, rpc_info:{Cop:{num_rpc:306, total_time:495.5ms}}, tikv_task:{proc max:6ms, min:0s, avg:597.3µs, p80:1ms, p95:1ms, iters:375, tasks:375}, scan_detail:{total_process_keys:400, total_process_keys_size:489856, total_keys:800, get_snapshot_time:158.3ms, rocksdb:{key_skipped_count:400, block:{cache_hit_count:3197, read_count:803, read_byte:10.2 MB, read_time:113.5ms}}}, time_detail:{total_process_time:220.7ms, total_suspend_time:5.39ms, total_wait_time:242.2ms, total_kv_read_wall_time:224ms, tikv_wall_time:430.5ms}} | keep order:false, stats:partial[...] | N/A | N/A |
```

The following sections describe similar detailed execution plans for partitioned tables with global and local indexes.

#### Create a global index on a partitioned table in TiDB

> There are two options for you to create a global index on a partitioned table in TiDB.

Reviewer: ```suggestion
You can create a global index on a partitioned table using one of the following methods.
```

> **Note:** 
>
> - In TiDB v8.5.3 and earlier versions, global indexes can only be created on unique columns. Starting from v8.5.4, global indexes on non-unique columns are supported. This limitation will be removed in the next LTS version.
> - For non-unique global indexes, use `ADD INDEX` instead of `ADD UNIQUE INDEX`.
> - The `GLOBAL` keyword must be explicitly specified.

##### Option 1: add via `ALTER TABLE`

> You can use `ALTER TABLE` to add a global index to an existing partitioned table.

Reviewer: ```suggestion
To add a global index to an existing partitioned table, use `ALTER TABLE`:
```

  1696271179344,
  1696317134004,
  1696181972136,
  ...
  1696159221765
);
```

This query pattern is representative because it:

- Filters on a secondary index without the partition key.
- Triggers a local index lookup for each partition due to lack of pruning.
- Generates significantly more table lookup tasks for partitioned tables.

#### Test results

The following table shows results for a query returning 400 rows from a table with 365 Range partitions.

| Configuration | Average query time | Cop tasks (index scan) | Cop tasks (table lookup) | Total Cop tasks |
|---|---|---|---|---|
| Non-partitioned table | 12.6 ms | 72 | 79 | 151 |
| Partitioned table with local indexes | 108 ms | 600 | 375 | 975 |
| Partitioned table with global indexes | 14.8 ms | 69 | 383 | 452 |

- **Non-partitioned table**: provides the best performance with the fewest tasks. Suitable for most OLTP workloads.
- **Partitioned table with global indexes**: improve index scan efficiency, but table lookups remain expensive when many rows match.
- **Partitioned table with local indexes**: when the query condition does not include the partition key, local index queries scan all partitions.

> 

Reviewer: ```suggestion

```

#### Create a global index on a partitioned table

You can create a global index on a partitioned table using one of the following methods.

> **Note:**
>
> - In TiDB v8.5.3 and earlier versions, you can only create global indexes on unique columns. Starting from v8.5.4, TiDB supports global indexes on non-unique columns. This limitation will be removed in a future LTS version.
> - For non-unique global indexes, use `ADD INDEX` instead of `ADD UNIQUE INDEX`.
> - You must explicitly specify the `GLOBAL` keyword.

##### Option 1: Use `ALTER TABLE`

To add a global index to an existing partitioned table, use `ALTER TABLE`:

```sql
ALTER TABLE <table_name>
ADD UNIQUE INDEX <index_name> (col1, col2) GLOBAL;
```

##### Option 2: Define the index at table creation

To create a global index when creating a table, define the global index inline in the `CREATE TABLE` statement:

```sql
CREATE TABLE t (
  id BIGINT NOT NULL,
  col1 VARCHAR(50),
  col2 VARCHAR(50),
  -- other columns...
  UNIQUE GLOBAL INDEX idx_col1_col2 (col1, col2)
)
PARTITION BY RANGE (id) (
  PARTITION p0 VALUES LESS THAN (10000),
  PARTITION p1 VALUES LESS THAN (20000),
  PARTITION pMax VALUES LESS THAN MAXVALUE
);
```

> #### Summary

Reviewer: ```suggestion
#### Performance summary
```

The performance overhead of partitioned tables in TiDB depends significantly on the number of partitions and the type of index used.

- The more partitions you have, the more severe the potential performance degradation.
- With a smaller number of partitions, the impact might not be noticeable, but it is still workload-dependent.
- For local indexes, if a query does not include effective partition pruning conditions, the number of partitions directly correlates with the number of [Remote Procedure Calls (RPCs)](https://docs.pingcap.com/tidb/stable/glossary/#remote-procedure-call-rpc) triggered. This means more partitions will likely result in more RPCs, leading to higher latency.

> - For global indexes, the number of RPCs and the degree of performance regression depend on both the number of partitions involved and how many rows need to be retrieved (that is, the number of rows requiring table lookups). Note that for very large tables where data is already distributed across many Regions, accessing data through a global index might have similar performance to a non-partitioned table, as both scenarios require multiple cross-Region RPCs.

Reviewer: ```suggestion
- **Global indexes**: the performance depends on both the number of partitions involved and the number of rows that require table lookups. For very large tables where data is distributed across multiple Regions, accessing data through a global index provides performance similar to that of a non-partitioned table, because both scenarios involve multiple cross-Region RPCs.
```

#### Recommendations

> 

Reviewer: ```suggestion

Use the following guidelines when you design partitioned tables and indexes in TiDB:

```

- Do not use partitioned tables unless necessary. For most OLTP workloads, a well-indexed non-partitioned table performs better and is easier to manage.
- If you are sure that all queries will make use of good partition pruning (matching only a few partitions), then local indexes are a good choice.
- If you are sure that critical queries do not have good partition pruning (matching many partitions), then a global index is recommended.
- Use local indexes only if your main concern is DDL efficiency (such as fast `DROP PARTITION`) and the performance side effect from the partition table is acceptable.

## Facilitate bulk data deletion

> In TiDB, you can clear up historical data either by TTL (Time-to-Live) or manual partition drop. While both methods serve the same purpose, they differ significantly in performance. The test cases in this section show that dropping partitions is generally faster and less resource-intensive, making it a better choice for large datasets and frequent purging needs.

Reviewer: ```suggestion
In TiDB, you can remove historical data by using [TTL (Time to Live)](/time-to-live.md) or by manually dropping partitions. Although both methods delete data, their performance characteristics differ significantly. The following test results show that dropping partitions is generally faster and consumes fewer resources, making it a better option for large datasets and frequent data purging.
```

### Differences between TTL and `DROP PARTITION`

- TTL: automatically removes data based on its age, but might be slower due to the need to scan and clean data over time.

> - `DROP PARTITION`: deletes an entire partition at once, making it much faster, especially when dealing with large datasets.

Reviewer: ```suggestion
- TTL: automatically deletes data based on its age. This method might be slower because it scans and deletes rows incrementally over time.
- `DROP PARTITION`: deletes an entire partition in a single operation. This approach is typically much faster, especially for large datasets.
```

#### Test case

> To compare the performance of TTL and `DROP PARTITION`, the test case in this section configures TTL to execute every 10 minutes and create a partitioned version of the same table, dropping one partition at the same interval for comparison. Both approaches are tested under background write workloads of 50 and 100 concurrent threads. This test case measures key metrics such as execution time, system resource utilization, and the total number of rows deleted.

Reviewer: ```suggestion
This test compares the performance of TTL and `DROP PARTITION`.

- TTL configuration: runs every 10 minutes.
- Partition configuration: drops one partition every 10 minutes.
- Workload: background write workloads with 50 and 100 concurrent threads.

The test measures execution time, system resource usage, and the total number of rows deleted.
```

#### Findings

> **Note:**
>
> The performance benefits described in this section only apply to partitioned tables without global indexes. 

The following are findings about the TTL performance:

- On a write-heavy table, TTL runs every 10 minutes.
- With 50 threads, each TTL job takes 8 to 10 minutes, deleting 7 to 11 million rows.
- With 100 threads, it handles up to 20 million rows, but the execution time increases to 15 to 30 minutes, with greater variance.

> - TTL jobs impact system performance under high workloads due to extra scanning and deletion activity, reducing overall QPS.

Reviewer: ```suggestion
- With 100 threads, TTL handles up to 20 million rows, but execution time increases to 15 to 30 minutes and shows higher variance.
- Under heavy workloads, TTL jobs reduce overall QPS due to additional scanning and deletion overhead.
```

The following are findings about the `DROP PARTITION` performance:

- `ALTER TABLE ... DROP PARTITION` removes an entire data segment instantly, with minimal resource usage.
- `ALTER TABLE ... DROP PARTITION` is a metadata-level operation, making it much faster and more predictable than TTL, especially when managing large volumes of historical data.

#### Use TTL and `DROP PARTITION` in TiDB

In this test case, the table structures have been anonymized. For more information about the usage of TTL, see [Periodically Delete Data Using TTL (Time to Live)](/time-to-live.md).

> The following is the TTL schema.

Reviewer: ```suggestion
The following example shows a TTL-enabled table schema:
```

INTERVAL (10 MINUTE)
FIRST PARTITION LESS THAN ('2025-02-19 18:00:00')
...
LAST PARTITION LESS THAN ('2025-02-19 20:00:00');
```

You need to run DDL statements such as `ALTER TABLE PARTITION ...` to change the `FIRST PARTITION` and `LAST PARTITION` periodically. These two DDL statements can drop the old partitions and create new ones.

```sql
ALTER TABLE ad_cache FIRST PARTITION LESS THAN ("${nextTimestamp}");
ALTER TABLE ad_cache LAST PARTITION LESS THAN ("${nextTimestamp}");
```

#### Recommendations

For workloads with large or time-based data cleanup, it is recommended to use partitioned tables with `DROP PARTITION`. It offers better performance, lower system impact, and simpler management. 

TTL is still useful for finer-grained or background cleanup, but might not be optimal under high write pressure or when deleting large volumes of data quickly.

### Partition drop efficiency: local indexes vs. global indexes

A partitioned table with a global index requires synchronous updates to the global index, which can significantly increase the execution time for DDL operations, such as `DROP PARTITION`, `TRUNCATE PARTITION`, and `REORGANIZE PARTITION`.

> In this section, the tests show that `DROP PARTITION` is much slower when using a global index compared to a local index. Take this into consideration when you design partitioned tables.

Reviewer: ```suggestion
This section shows that `DROP PARTITION` is substantially slower on tables with global indexes than on tables with local indexes. Consider this behavior when you design partitioned tables.
```

#### Test case

This test case creates a table with 365 partitions and tests the `DROP PARTITION` performance using both global indexes and local indexes. The total number of rows is 1 billion.

> | Index type   | Duration (drop partition) |

Reviewer: ```suggestion
| Index type   | Drop partition duration |
```

  1696271179344,
  1696317134004,
  1696181972136,
  ...
  1696159221765
);
```

This query pattern is representative because it:

- Filters on a secondary index without the partition key.
- Triggers a local index lookup for each partition due to lack of pruning.
- Generates significantly more table lookup tasks for partitioned tables.

#### Test results

The following table shows results for a query returning 400 rows from a table with 365 Range partitions.

| Configuration | Average query time | Cop tasks (index scan) | Cop tasks (table lookup) | Total Cop tasks |
|---|---|---|---|---|
| Non-partitioned table | 12.6 ms | 72 | 79 | 151 |
| Partitioned table with local indexes | 108 ms | 600 | 375 | 975 |
| Partitioned table with global indexes | 14.8 ms | 69 | 383 | 452 |

- **Non-partitioned table**: provides the best performance with the fewest tasks. Suitable for most OLTP workloads.
- **Partitioned table with global indexes**: improve index scan efficiency, but table lookups remain expensive when many rows match.
- **Partitioned table with local indexes**: when the query condition does not include the partition key, local index queries scan all partitions.

> > **Note:**

Reviewer: ```suggestion

> **Note:**
```

- Limited performance gains from scaling out TiKV nodes, as the bottleneck remains concentrated on one Region.

Partitioned tables can help mitigate this problem. By applying hash or key partitioning on the primary key, TiDB can spread inserts across multiple partitions (and therefore multiple Regions), reducing hotspot contention.

> **Note:** 
>
> This section uses partitioned tables as an example for mitigating read and write hotspots. TiDB also provides other features such as [`AUTO_INCREMENT`](/auto-increment.md) and [`SHARD_ROW_ID_BITS`](/shard-row-id-bits.md) for hotspot mitigation. When using partitioned tables in certain scenarios, you might need to set `merge_option=deny` to maintain partition boundaries. For more details, see [issue #58128](https://github.com/pingcap/tidb/issues/58128).

### How it works

TiDB stores table data and indexes in Regions, each covering a continuous range of row keys. When the primary key is [`AUTO_INCREMENT`](/auto-increment.md) and the secondary indexes on datetime columns are monotonically increasing:

**Without partitioning:**

- New rows always have the highest key values and are inserted into the same "last Region."

> - That Region is served by one TiKV node at a time, becoming a single write bottleneck.

Reviewer: ```suggestion
**Non-partitioned tables**

In a non-partitioned table, new rows always have the largest key values and are written to the same "last" Region. This single Region, served by one TiKV node, can become a write bottleneck.
```

> **With hash or key partitioning:**

Reviewer: ```suggestion
**Hash or Key partitioned tables**
```

- The table and the secondary indexes are split into multiple partitions using a hash or key function on the primary key or indexed columns.
- Each partition has its own set of Regions, often distributed across different TiKV nodes.

> - Inserts are spread across multiple Regions in parallel, improving workload distribution and throughput.

Reviewer: ```suggestion
- TiDB splits the table and its indexes into multiple partitions by applying a Hash or Key function to the primary key or indexed columns.
- Each partition has its own set of Regions, which are typically distributed across different TiKV nodes.
- Insert operations are distributed across multiple Regions in parallel, improving workload balance and write throughput.
```

### Use cases

> If a table with an [`AUTO_INCREMENT`](/auto-increment.md) primary key experiences heavy bulk inserts and suffers from write hotspot issues, applying hash or key partitioning on the primary key can help distribute the write workload more evenly.

Reviewer: ```suggestion
If a table with an [`AUTO_INCREMENT`](/auto-increment.md) primary key receives heavy bulk inserts and experiences write hotspots, apply Hash or Key partitioning to the primary key to distribute the write workload more evenly.
```

```sql
CREATE TABLE server_info (
  id bigint NOT NULL AUTO_INCREMENT,
  serial_no varchar(100) DEFAULT NULL,
  device_name varchar(256) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci DEFAULT NULL,
  device_type varchar(50) DEFAULT NULL,
  modified_ts timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
  PRIMARY KEY (id) /*T![clustered_index] CLUSTERED */,
  KEY idx_serial_no (serial_no),
  KEY idx_modified_ts (modified_ts)
) /*T![auto_id_cache] AUTO_ID_CACHE=1 */
PARTITION BY KEY (id) PARTITIONS 16;
```

### Advantages

> 

Reviewer: ```suggestion
### Benefits

Partitioned tables provide the following benefits:

```

- Balanced write workload: hotspots are spread across multiple partitions, and therefore multiple Regions, reducing contention and improving insert performance.
- Query optimization via partition pruning: if queries already filter by the partition key, TiDB can prune unused partitions, scanning less data and improving query speed.

### Disadvantages

There are some risks when using partitioned tables.

- When converting a non-partitioned table to a partitioned table, TiDB creates separate Regions for each partition. This might significantly increase the total Region count. Queries that do not filter by the partition key cannot take advantage of partition pruning, forcing TiDB to scan all partitions or do index lookups in all partitions. This increases the number of coprocessor (cop) tasks and can slow down queries. For example, `serial_no` is not the partition key, which will cause the query performance regression:

    ```sql
    SELECT * FROM server_info WHERE `serial_no` = ?;
    ```

> - Add a global index on the filtering columns used by these queries to reduce scanning overhead. While creating a global index can significantly slow down `DROP PARTITION` operations, hash and key partitioned tables do not support `DROP PARTITION`. In practice, such partitions are rarely truncated, making global indexes a feasible solution in these scenarios. For example:

Reviewer: ```suggestion
- To reduce scan overhead for queries that do not use the partition key, you need to create a global index. Although global indexes can slow down `DROP PARTITION` operations, Hash and Key partitioned tables do not support `DROP PARTITION`. Therefore, global indexes are a practical solution because these partitions are rarely truncated. For example:
```

    ```sql
    ALTER TABLE server_info ADD UNIQUE INDEX(serial_no, id) GLOBAL;
    ```

## Partition management challenges

> New range partitions in a partitioned table can easily lead to hotspot issues in TiDB. This section outlines common scenarios and mitigation strategies to avoid read and write hotspots caused by new range partitions.

Reviewer: ```suggestion
New Range partitions can cause hotspot issues in TiDB. This section describes common scenarios and provides mitigation strategies.
```

### Read hotspots

When using range-partitioned tables, if queries do not filter data using the partition key, new empty partitions can easily become read hotspots.

**Root cause:**

> By default, TiDB creates an empty region for each partition when the table is created. If no data is written for a while, multiple empty partitions' regions might be merged into a single region.

Reviewer: ```suggestion
By default, TiDB creates an empty Region for each partition when you create the table. If no data is written for a period, TiDB might merge Regions for multiple empty partitions into a single Region.
```

> **impact:**

Reviewer: ```suggestion
**Impact:**
```

When a query does not filter by partition key, TiDB scans all partitions (as seen in the execution plan `partition:all`). As a result, the single region holding multiple empty partitions is scanned repeatedly, leading to a read hotspot.

### Write hotspots

When using a time-based field as the partition key, a write hotspot might occur when switching to a new partition.

**Root cause:**

> In TiDB, newly created partitions initially contain only one region on a single TiKV node. As writes concentrate on this single region, it must split into multiple regions before writes can be distributed across multiple TiKV nodes. This splitting process is the main cause of the temporary write hotspot. 

Reviewer: ```suggestion
In TiDB, newly created partitions initially contain a single Region on one TiKV node. All writes are directed to this single Region until it splits and data redistributes. During this period, the TiKV node must handle both application writes and Region-splitting tasks.
```

> However, if the initial write traffic to this new partition is very high, the TiKV node hosting that single initial region will be under heavy write pressure. In such cases, it might not have enough spare resources (such as I/O capacity and CPU cycles) to handle both the application writes and the scheduling of newly split regions to other TiKV nodes. This can delay region distribution, keeping most writes concentrated on the same node for longer than desired.

Reviewer: ```suggestion
If the initial write traffic to the new partition is very high, the TiKV node might not have sufficient resources (such as CPU or I/O capacity) to split and scatter Regions promptly. As a result, writes remain concentrated on the same node longer than expected.
```

**Impact:**

> This imbalance can cause the TiKV node to trigger flow control, leading to a sharp drop in QPS, a spike in write latency, and increased CPU usage on the affected node, which in turn might impact the overall read and write performance of the cluster.

Reviewer: ```suggestion
This imbalance can trigger flow control on the TiKV node, leading to a sharp drop in QPS, increased write latency, and high CPU utilization, which can degrade overall cluster performance.
```

> ### Summary

Reviewer: ```suggestion
### Comparison of partitioned table types
```

The following table shows the summary information for non-clustered and clustered partitioned tables.

> | Table type                      | Region pre-splitting | Read performance     | Write scalability | Data cleanup via partition |

Reviewer: ```suggestion
| Table type                      | Region pre-splitting | Read performance     | Write scalability | Data cleanup by partition |
```

|---|---|---|---|---|
| Non-clustered partitioned table | Automatic            | Lower (more lookups) | High              | Supported          | 

> | Clustered partitioned table     | Manual               | High (fewer lookups) | High (if managed) | Supported          | 

Reviewer: ```suggestion
| Non-clustered partitioned table | Automatic            | Lower (additional lookups required) | High              | Supported          | 
| Clustered partitioned table     | Manual               | High (fewer lookups) | High (with manual management) | Supported          | 
```

| Clustered non-partitioned table | N/A                  | High                 | Stable            | Not supported      |

### Solutions for non-clustered partitioned tables

#### Advantages

- When a new partition is created in a non-clustered partitioned table configured with [`SHARD_ROW_ID_BITS`](/shard-row-id-bits.md) and [`PRE_SPLIT_REGIONS`](/sql-statements/sql-statement-split-region.md#pre_split_regions), the regions can be automatically pre-split, significantly reducing manual intervention.
- Lower operational overhead.

#### Disadvantages

> Queries using **Point Get** or **Table Range Scan** require more table lookups, which can degrade read performance for such query types.

Reviewer: ```suggestion
Queries using **Point Get** or **Table Range Scan** require additional table lookups, which can degrade read performance.
```

ALTER TABLE employees ATTRIBUTES 'merge_option=deny';
-- partition
ALTER TABLE employees PARTITION `p3` ATTRIBUTES 'merge_option=deny';
```

##### Step 3. Determine split boundaries based on existing business data

To avoid hotspots when a new table or partition is created, it is often beneficial to pre-split regions before heavy writes begin. To make pre-splitting effective, configure the lower and upper boundaries for region splitting based on the actual business data distribution. Avoid setting excessively wide boundaries, as this can result in real data not being effectively distributed across TiKV nodes, defeating the purpose of pre-splitting.

Identify the minimum and maximum values from existing production data so that incoming writes are more likely to target different pre-allocated regions. The following is an example query for the existing data:

```sql
SELECT MIN(id), MAX(id) FROM employees;
```

- If the table is new and has no historical data, estimate the minimum and maximum values based on your business logic and expected data range.
- For composite primary keys or composite indexes, only the leftmost column needs to be considered when deciding split boundaries.

> - If the leftmost column is a string, take string length and distribution into account to ensure even data spread.

Reviewer: ```suggestion
- If the table has no historical data, estimate the minimum and maximum values based on business requirements and expected data ranges.
- For composite primary keys or composite indexes, use only the leftmost column to define split boundaries.
- If the leftmost column is a string, consider its length and value distribution to ensure even data distribution.
```

##### Step 4. Pre-split and scatter regions

A common practice is to split the number of regions to match the number of TiKV nodes, or to be twice the number of TiKV nodes. This helps ensure that data is more evenly distributed across the cluster from the start.

> ##### Step 5. Split regions for the primary key and the secondary index of all partitions if needed

Reviewer: ```suggestion
##### Step 5. Split Regions for primary and secondary indexes if needed
```

To split regions for the primary key of all partitions in a partitioned table, use the following SQL statement:

```sql
SPLIT PARTITION TABLE employees INDEX `PRIMARY` BETWEEN (1, "1970-01-01") AND (100000, "9999-12-31") REGIONS <number_of_regions>;
```

> This example splits each partition's primary key range into `<number_of_regions>` regions between the specified boundary values.

Reviewer: ```suggestion
This example splits each partition's primary key range into `<number_of_regions>` Regions within the specified boundaries.
```

```sql
SPLIT PARTITION TABLE employees INDEX `idx_employees_on_store_id` BETWEEN (1) AND (1000) REGIONS <number_of_regions>;
```

##### Step 6. (Optional) When adding a new partition, you need to manually split regions for its primary key and indexes

```sql
ALTER TABLE employees ADD PARTITION (PARTITION p4 VALUES LESS THAN (2011));

SHOW TABLE employees PARTITION (p4) regions;

SPLIT PARTITION TABLE employees INDEX `PRIMARY` BETWEEN (1, "2006-01-01") AND (100000, "2011-01-01") REGIONS <number_of_regions>;

SPLIT PARTITION TABLE employees PARTITION (p4) INDEX `idx_employees_on_store_id` BETWEEN (1) AND (1000) REGIONS <number_of_regions>;

SHOW TABLE employees PARTITION (p4) regions;
```

### Solutions for clustered partitioned tables

#### Advantages

Queries using **Point Get** or **Table Range Scan** do not need additional lookups, resulting in better read performance.

#### Disadvantages

Manual region splitting is required when creating new partitions, increasing operational complexity.

#### Suitable scenarios

It is suitable when low-latency point queries are important and operational resources are available to manage region splitting.

#### Best practices

> To address hotspot issues caused by new range partitions, you can perform the steps described in [Best practices for non-clustered partitioned tables](#best-practices).

Reviewer: ```suggestion
To mitigate hotspot issues caused by new Range partitions, follow the steps in [Best practices for non-clustered partitioned tables](#best-practices).
```

### Solutions for clustered non-partitioned tables

#### Advantages

- No hotspot risks from new range partitions.
- Provides good read performance for point and range queries.

#### Disadvantages

You cannot use `DROP PARTITION` to clean up large volumes of old data to improve deletion efficiency.

#### Suitable scenarios

It is suitable for scenarios that require stable performance and do not benefit from partition-based data management.

## Convert between partitioned and non-partitioned tables

> When working with large tables (for example, a table with 120 million rows), transforming between partitioned and non-partitioned schemas is sometimes required for performance tuning or schema design changes. TiDB supports several main approaches for such transformations:

Reviewer: ```suggestion
For large tables, such as those with 120 million rows, you might need to convert between partitioned and non-partitioned schemas for performance tuning or schema redesign. TiDB supports the following approaches:
```

- [Pipelined DML](/pipelined-dml.md): `INSERT INTO ... SELECT ...`
- [`IMPORT INTO`](/sql-statements/sql-statement-import-into.md): `IMPORT INTO ... FROM SELECT ...`

> - [Online DDL](/dm/feature-online-ddl.md): Direct schema transformation via `ALTER TABLE`

Reviewer: ```suggestion
- [Online DDL](/dm/feature-online-ddl.md): direct schema transformation using `ALTER TABLE`
```

```sql
CREATE TABLE `fa` (
  `id` bigint NOT NULL AUTO_INCREMENT,
  `account_id` bigint(20) NOT NULL,
  `sid` bigint(20) DEFAULT NULL,
  `user_id` bigint NOT NULL,
  `date` int NOT NULL,
  PRIMARY KEY (`id`,`date`) /*T![clustered_index] CLUSTERED */,
  KEY `index_fa_on_sid` (`sid`),
  KEY `index_fa_on_account_id` (`account_id`),
  KEY `index_fa_on_user_id` (`user_id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
PARTITION BY RANGE (`date`)
(PARTITION `fa_2024001` VALUES LESS THAN (2025001),
PARTITION `fa_2024002` VALUES LESS THAN (2025002),
PARTITION `fa_2024003` VALUES LESS THAN (2025003),
...
...
PARTITION `fa_2024365` VALUES LESS THAN (2025365));
```

### Table schema for a non-partitioned table: `fa_new`

```sql
CREATE TABLE `fa_new` (
  `id` bigint NOT NULL AUTO_INCREMENT,
  `account_id` bigint(20) NOT NULL,
  `sid` bigint(20) DEFAULT NULL,
  `user_id` bigint NOT NULL,
  `date` int NOT NULL,
  PRIMARY KEY (`id`,`date`) /*T![clustered_index] CLUSTERED */,
  KEY `index_fa_on_sid` (`sid`),
  KEY `index_fa_on_account_id` (`account_id`),
  KEY `index_fa_on_user_id` (`user_id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
```

> These examples show converting a partitioned table to a non-partitioned table, but the same methods also work for converting a non-partitioned table to a partitioned table.

Reviewer: ```suggestion
These examples demonstrate converting a partitioned table to a non-partitioned table. The same methods apply when converting a non-partitioned table to a partitioned table.
```

INSERT INTO fa_new SELECT * FROM fa;
-- 120 million rows copied in 58m 42s
```

### Method 2: `IMPORT INTO ... FROM SELECT`

```sql
IMPORT INTO fa_new FROM SELECT * FROM fa WITH thread = 32, disable_precheck;
```

```
Query OK, 120000000 rows affected, 1 warning (16 min 49.90 sec)
Records: 120000000, ID: c1d04eec-fb49-49bb-af92-bf3d6e2d3d87
```

### Method 3: Online DDL

The following SQL statement converts from a partition table to a non-partitioned table:

```sql
SET @@global.tidb_ddl_REORGANIZE_worker_cnt = 16;
SET @@global.tidb_ddl_REORGANIZE_batch_size = 4096;
ALTER TABLE fa REMOVE PARTITIONING;
-- real 170m12.024 s (≈ 2 h 50 m)
```

The following SQL statement converts from a non-partition table to a partitioned table:

```sql
SET @@global.tidb_ddl_REORGANIZE_worker_cnt = 16;
SET @@global.tidb_ddl_REORGANIZE_batch_size = 4096;
ALTER TABLE fa_new PARTITION BY RANGE (`date`)
(PARTITION `fa_2024001` VALUES LESS THAN (2025001),
PARTITION `fa_2024002` VALUES LESS THAN (2025002),
...
PARTITION `fa_2024365` VALUES LESS THAN (2025365),
PARTITION `fa_2024366` VALUES LESS THAN (2025366));

Query OK, 0 rows affected, 1 warning (2 hours 31 min 57.05 sec)
```

### Findings

> The following table show the time taken by each method.

Reviewer: ```suggestion
The following table shows the time taken by each method for a 120-million-row table:
```
